### PR TITLE
implement user account page

### LINKED ISSUE #44: Implement "User Account" page
In order to provide user with a way to edit their account information - we need to create separate page, which will be accessed through User menu in the application header.

This page should allow to user to do following changes:

- Change name, email and password
- Upload avatar
- Delete account

This also means we need to implement proper API routes to handle this functionality

### PR DESCRIPTION
**Pull request by Artemh59**
Closes #44

### CODE REVIEW

File: src/app/api/user/[id]/route.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+};
+
+export async function PUT(request: Request, { params }: { params: { id: string } }) {
+  const session = await auth();
+  const userId = session?.user?.id;
+
+  try {
+    const id = params.id;
+
+    if (!userId) {
```
Comment (mudrila) [inline src/app/api/user/[id]/route.ts:22]: Think this might be even before `try..catch` block

File: src/app/api/user/[id]/route.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+import prisma from '@/utils/prisma';
+
+type UpdateData = {
+  name?: string;
+  email?: string;
+  password?: string;
+  image?: string;
+};
+
+export async function PUT(request: Request, { params }: { params: { id: string } }) {
```
Comment (mudrila) [inline src/app/api/user/[id]/route.ts:15]: Oh, I just realized - we don't need to put it under `/api/user/[id]` - could be just `/api/account` and working with `userId` that is supplied from session, no need for `params.id`

File: src/app/api/user/[id]/route.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+    if (password) {
+      if (!password.match(passwordRegex)) {
+        return NextResponse.json({ errorMessage: 'Invalid password format' });
+      }
+      const passwordHash = await generateHashPassword(password);
+      updateData.password = passwordHash.hash;
+    }
+
+    if (Object.keys(updateData).length === 0) {
```
Comment (mudrila) [inline src/app/api/user/[id]/route.ts:57]: Let's put it before very first `if(name)` - `if (!name || !email || !password || !image) ... `

File: src/app/app/user/page.tsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { Metadata } from 'next';
+import Profile from '@/components/pages/User/Profile';
+import { APP_SHORT_NAME } from '@/constants/content';
+import { auth } from '@/utils/auth';
+
+export const metadata: Metadata = {
+  title: `User page | ${APP_SHORT_NAME}`,
```
Comment (mudrila) [inline src/app/app/user/page.tsx:7]: Let's call it `Your Account` instead and move it to `src/app/app/account`

File: src/components/molecules/Header/Header.tsx [hunk 5]
```diff
@@ -82,9 +82,7 @@ export default function Header() {
           open={Boolean(anchorElUser)}
           onClose={handleCloseUserMenu}
         >
-          <Typography sx={{ textAlign: 'center', p: 1 }}>
-            Hello, {session.data?.user?.name}
-          </Typography>
+          <StyledLink href="/app/user">Hello, {session.data?.user?.name}</StyledLink>
```
Comment (mudrila) [inline src/components/molecules/Header/Header.tsx:85]: Let's change it to `MenuItem`, put there "manage account icon"

```
import ManageAccountsIcon from '@mui/icons-material/ManageAccounts';
```

File: src/components/pages/User/Profile.tsx [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+import { toast } from 'sonner';
+import EmailInput from '@/components/molecules/Inputs/EmailInput/EmailInput';
+import PasswordInput from '@/components/molecules/Inputs/PasswordInput/PasswordInput';
+
+interface ProfileProps {
+  userName: string;
+  userId: string;
+}
+
+export default function Profile({ userName, userId }: ProfileProps) {
```
Comment (mudrila) [inline src/components/pages/User/Profile.tsx:27]: Let's call it `Account`

**Review by mudrila — CHANGES_REQUESTED** [review]
Few minor changes but should be good to go then

File: src/app/api/user/[id]/route.ts (hunk 1 above)
Comment (Artemh59) [reply src/app/api/user/[id]/route.ts:22]: Do I need to rewrite it everywhere in this commit?

**Comment by Artemh59** [comment]
and when we change some user's information, we don't see it on page, to see this we need to logout and again login, but in database everything change on time, how we can fix it?

File: src/app/api/user/[id]/route.ts (hunk 1 above)
Comment (mudrila) [reply src/app/api/user/[id]/route.ts]: Yes please

**Comment by mudrila** [comment]
> and when we change some user's information, we don't see it on page, to see this we need to logout and again login, but in database everything change on time, how we can fix it?

So, for that - we'll need to adjust the way we're reading / writing to `jwt` and `session` on initializing `next-auth`. I do have a clear picture for fixing / implementing this - but let's tackle it later

File: src/app/app/account/page.tsx [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+export const metadata: Metadata = {
+  title: `Your Account | ${APP_SHORT_NAME}`,
+};
+
+export default async function UserPage() {
+  const session = await auth();
+  const userName = session?.user?.name;
+
+  if (userName) {
+    return <Profile userName={userName} />;
```
Comment (mudrila) [inline src/app/app/account/page.tsx:15]: Well, given we do require `userName` during profile creation - it is safe to assume that if it's not set in session - `user` is not authorized at all. Though user won't get here due to `middleware.ts` validation - to provide better type compatibility and logic handling - let's flip this condition 
```
if (session?.user) {
  redirect('/sign-in')
}

return <Profile userName={userName}
```

File: src/components/pages/User/Account.tsx [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+import { toast } from 'sonner';
+import EmailInput from '@/components/molecules/Inputs/EmailInput/EmailInput';
+import PasswordInput from '@/components/molecules/Inputs/PasswordInput/PasswordInput';
+
+interface ProfileProps {
+  userName: string;
+}
+
+export default function Account({ userName }: ProfileProps) {
+  const [image, setImage] = useState<File>();
```
Comment (mudrila) [inline src/components/pages/User/Account.tsx:27]: We should pre-fill it from user's session information

File: src/components/pages/User/Account.tsx [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+  const [base64Image, setBase64Image] = useState('');
+  const [openAvatar, setOpenAvatar] = useState(false);
+  const [openDelete, setOpenDelete] = useState(false);
+
+  const [name, setName] = useState('');
```
Comment (mudrila) [inline src/components/pages/User/Account.tsx:32]: And here

File: src/components/pages/User/Account.tsx [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+  const [email, setEmail] = useState('');
```
Comment (mudrila) [inline src/components/pages/User/Account.tsx:33]: And here

**Review by mudrila — CHANGES_REQUESTED** [review]
Bunch of changes, but should be good to go then!

**Review by mudrila — APPROVED** [review]

**Pull request merged**